src/main/scala/diplomacy/LazyModule.scala

```diff
@@ -144,52 +215,79 @@ sealed trait LazyModuleImpLike extends RawModule
   val auto: AutoBundle
   protected[diplomacy] val dangles: Seq[Dangle]
 
-  // .module had better not be accessed while LazyModules are still being built!
-  require (!LazyModule.scope.isDefined, s"${wrapper.name}.module was constructed before LazyModule() was run on ${LazyModule.scope.get.name}")
+  /** [[wrapper.module]] had better not be accessed while LazyModules are still being built! */
+  require (LazyModule.scope.isEmpty, s"${wrapper.name}.module was constructed before LazyModule() was run on ${LazyModule.scope.get.name}")
 
+  /** set name from [[wrapper]]. */
   override def desiredName = wrapper.desiredName
   suggestName(wrapper.suggestedName)
 
-  implicit val p = wrapper.p
+  /** [[Parameters]] for diplomatic [[Module]]s. */
+  implicit val p: Parameters = wrapper.p
 
+  /** [[instantiate]] will be called when a instance of [[LazyModuleImp]] is created. */
   protected[diplomacy] def instantiate() = {
     val childDangles = wrapper.children.reverse.flatMap { c =>
       implicit val sourceInfo = c.info
+      /** calling [[c.module]] will push the real [[Module]] into [[chisel3.internal.Builder]],
+        * notice this place is not calling [[instantiate]], it just push the [[c.module]] to Builder.
+        * */
       val mod = Module(c.module)
+      /** ask each child to finish instantiate. */
       mod.finishInstantiate()
+      /** return dangles of each child. */
       mod.dangles
     }
+
+    /** ask each node in the [[LazyModule]] to call [[BaseNode.instantiate]].
+      * it will return a sequence of [[Dangle]] of these [[BaseNode]].
+      * */
     val nodeDangles = wrapper.nodes.reverse.flatMap(_.instantiate())
+    /** add all node and child dangle together. */
     val allDangles = nodeDangles ++ childDangles
+    /** internal nodes pairing, group [[allDangles]] which has same [[Dangle.source]]*/
     val pairing = SortedMap(allDangles.groupBy(_.source).toSeq:_*)
+    /** connect dangles can be paired. */
```

Comment:
```suggestion
    /** For each set of [[Dangle]]s where there are sets of 2 with the same source, ensure that they can be connected in a source/sink pair, and then make the connection. For these sets, mark them as [[done]].*/
```

I don't understand what this iss doing really though. If they have the same source, why would they connect together...? I feel like an example here would really help.

Reply:
OK, I'll give a example later.

Reply:
They have a same source, but different `flipped` required.
```
      require (a.flipped != b.flipped)
```
This is because the `Dangle` generated in Nodes are always from source to sink. `inwardNode` and `outwardNode` has a sample `source` and `sink` but different `flipped`